### PR TITLE
Add server-side graphQL client

### PR DESCRIPTION
**Pull request by MadeByMike**
Added a new package `@keystonejs/server-side-graphql-client` a library for running server-side graphQL queries and mutations in Keystone. 

@timleslie I really want to consider changing `extraContext` to simply, `context` so that it replaces rather than extends the existing context inline with the `executeGraphQL` method.

### CODE REVIEW

File: packages/server-side-graphql-client/README.md [hunk 1]
```diff
@@ -0,0 +1,392 @@
+<!--[meta]
+section: api
+title: Server-side graphQL client
```
Comment (timleslie) [inline packages/server-side-graphql-client/README.md:3]:
```suggestion
title: Server-side GraphQL Client
```

File: packages/server-side-graphql-client/README.md [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+[meta]-->
+
+# Server-side graphQL client
```
Comment (timleslie) [inline packages/server-side-graphql-client/README.md:6]:
```suggestion
# Server-side GraphQL Client
```

File: packages/server-side-graphql-client/README.md [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+
+A library for running server-side graphQL queries and mutations in Keystone. It replaces the `keystone.createItems` method with a set of utility functions to generate and execute graphQL queries.
```
Comment (timleslie) [inline packages/server-side-graphql-client/README.md:8]:
```suggestion
A library for running server-side graphQL queries and mutations in Keystone.
```

Let's not leave references to `createItems` in here. It's really weird to have docs reference historical things that no longer exist, particularly for new users who have never seen (and will never see) `keystone.createItems`. Let's describe this package in its own right.

File: packages/server-side-graphql-client/package.json [hunk 4]
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "@keystonejs/server-side-graphql-client",
+  "description": "A library for running server-side graphQL queries and mutations against in Keystone.",
```
Comment (timleslie) [inline packages/server-side-graphql-client/package.json:3]: "against in Keystone" ?

File: packages/server-side-graphql-client/tests/index.test.js [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+} = require('../lib/server-side-graphql-client');
+
+const testData = [
+  {
+    data: {
+      name: 'test',
+      age: 30,
+    },
+  },
+  { data: { name: 'test2', age: 40 } },
```
Comment (timleslie) [inline packages/server-side-graphql-client/tests/index.test.js:22]: Looks weird having one of these as a one-liner and the other spread over 4 lines.

File: packages/server-side-graphql-client/tests/index.test.js [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+      });
+    },
+  });
+}
+
+multiAdapterRunners().map(({ runner, adapterName }) =>
+  describe(`Adapter: ${adapterName}`, () => {
+    describe('create', () => {
+      test(
+        'Should create and get single item',
```
Comment (timleslie) [inline packages/server-side-graphql-client/tests/index.test.js:52]:
```suggestion
        'createItem, getItem: Should create and get single item',
```

Let's include the name of the function under test in the test name. Very helpful when things go wrong.

File: .changeset/two-dogs-push.md [hunk 7]
```diff
@@ -0,0 +1,29 @@
+---
+'@keystonejs/server-side-graphql-client': major
+---
+
+Added a new package `@keystonejs/server-side-graphql-client` a library for running server-side graphQL queries and mutations in Keystone.
```
Comment (timleslie) [inline .changeset/two-dogs-push.md:5]:
```suggestion
This is the initial release of `@keystonejs/server-side-graphql-client,` a library for running server-side graphQL queries and mutations in Keystone. 
```

File: packages/server-side-graphql-client/README.md [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+For custom queries use [`runCustomQuery`](#runcustomquery).
+
+- `createItem`
+- `createItems`
+- `updateItem`
+- `updateItems`
+- `deleteItem`
+- `deleteItems`
+
+For custom queries use `runCustomQuery`
```
Comment (timleslie) [inline packages/server-side-graphql-client/README.md:72]:
```suggestion
```

File: packages/server-side-graphql-client/README.md [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+
+> NOTE: All functions accepts a config object as an argument, and return a `Promise`.
```
Comment (timleslie) [inline packages/server-side-graphql-client/README.md:74]:
```suggestion
> NOTE: All functions accept a config object as an argument, and return a `Promise`.
```

File: packages/server-side-graphql-client/README.md [hunk 10]
```diff
@@ -0,0 +1,392 @@
...
+addUser({ name: 'keystone user', email: 'keystone@test.com' });
+```
+
+#### Config
+
+[Shared Config Options](#shared-config-options) apply to this function.
+
+| Properties | Type                            | Default    | Description              |
+| ---------- | ------------------------------- | ---------- | ------------------------ |
+| `item`     | GraphQL `[listName]CreateInput` | (required) | The item to be inserted. |
```
Comment (timleslie) [inline packages/server-side-graphql-client/README.md:123]:
```suggestion
| `item`     | GraphQL `[listName]CreateInput` | (required) | The item to be created. |
```

File: packages/server-side-graphql-client/README.md [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+addUsers();
+```
+
+#### Config
+
+[Shared Config Options](#shared-config-options) apply to this function.
+
+| Properties | Type                             | Default    | Description                                                                                    |
+| ---------- | -------------------------------- | ---------- | ---------------------------------------------------------------------------------------------- |
+| `items`    | GraphQL `[listName]sCreateInput` | (required) | The array of objects to be inserted.                                                           |
```
Comment (timleslie) [inline packages/server-side-graphql-client/README.md:166]:
```suggestion
| `items`    | GraphQL `[listName]sCreateInput` | (required) | The array of objects to be created.                                                           |
```

File: packages/server-side-graphql-client/README.md [hunk 12]
```diff
@@ -0,0 +1,392 @@
...
+| `pageSize` | `Number`                         | 500        | The create mutation batch size. This is useful when you have large set of data to be inserted. |
+
+### getItem
+
+Retrieve single item by its id.
```
Comment (timleslie) [inline packages/server-side-graphql-client/README.md:171]:
```suggestion
Retrieve a single item by its ID.
```

File: packages/server-side-graphql-client/README.md [hunk 13]
```diff
@@ -0,0 +1,392 @@
...
+[Shared Config Options](#shared-config-options) apply to this function.
+
+| Properties | Type                           | Default | Description                                                                                                |
+| ---------- | ------------------------------ | ------- | ---------------------------------------------------------------------------------------------------------- |
+| `where`    | GraphQL `[listName]WhereInput` | `{}`    | Limit results to items matching [where clause](https://www.keystonejs.com/guides/intro-to-graphql/#where). |
+| `pageSize` | `Number`                       | 500     | The query batch size. Useful when retrieving a large set of data.                                          |
+
+### updateItem
+
+Update single item.
```
Comment (timleslie) [inline packages/server-side-graphql-client/README.md:246]:
```suggestion
Update a single item.
```

File: packages/server-side-graphql-client/README.md [hunk 14]
```diff
@@ -0,0 +1,392 @@
...
+
+[Shared Config Options](#shared-config-options) apply to this function.
+
+| Properties | Type                            | Default    | Description             |
+| ---------- | ------------------------------- | ---------- | ----------------------- |
+| `item`     | GraphQL `[listName]UpdateInput` | (required) | The item to be updated. |
+
+### updateItems
+
+Allow bulk updating of items.
```
Comment (timleslie) [inline packages/server-side-graphql-client/README.md:282]:
```suggestion
Update multiple items.
```

File: packages/server-side-graphql-client/README.md [hunk 15]
```diff
@@ -0,0 +1,392 @@
...
+
+[Shared Config Options](#shared-config-options) apply to this function.
+
+| Properties | Type                             | Default    | Description                   |
+| ---------- | -------------------------------- | ---------- | ----------------------------- |
+| `items`    | GraphQL `[listName]sUpdateInput` | (required) | Array of items to be updated. |
+
+### deleteItem
+
+Delete single item.
```
Comment (timleslie) [inline packages/server-side-graphql-client/README.md:324]:
```suggestion
Delete a single item.
```

File: packages/server-side-graphql-client/README.md [hunk 16]
```diff
@@ -0,0 +1,392 @@
...
+
+[Shared Config Options](#shared-config-options) apply to this function.
+
+| Properties | Type     | Default    | Description                         |
+| ---------- | -------- | ---------- | ----------------------------------- |
+| `itemId`   | `String` | (required) | The `id` of the item to be deleted. |
+
+### deleteItems
+
+Allow bulk deleting of items.
```
Comment (timleslie) [inline packages/server-side-graphql-client/README.md:355]:
```suggestion
Delete multiple items.
```

File: packages/server-side-graphql-client/README.md [hunk 17]
```diff
@@ -0,0 +1,392 @@
...
+
+[Shared Config Options](#shared-config-options) apply to this function.
+
+| Properties | Type       | Default    | Description                        |
+| ---------- | ---------- | ---------- | ---------------------------------- |
+| `itemId`   | `String[]` | (required) | Array of item `id`s to be deleted. |
+
+### runCustomQuery
+
+Allow executing a custom query.
```
Comment (timleslie) [inline packages/server-side-graphql-client/README.md:386]:
```suggestion
Execute a custom query.
```

File: packages/server-side-graphql-client/README.md (hunk 1 above)
Comment (MadeByMike) [reply packages/server-side-graphql-client/README.md:3]: Nope, all our titles are sentence case

Comment (MadeByMike) [reply packages/server-side-graphql-client/README.md:3]: This is used in the Nav actually

File: packages/server-side-graphql-client/README.md (hunk 2 above)
Comment (MadeByMike) [reply packages/server-side-graphql-client/README.md:6]: Same as above

File: packages/server-side-graphql-client/README.md (hunk 3 above)
Comment (MadeByMike) [reply packages/server-side-graphql-client/README.md:8]: ok

File: packages/server-side-graphql-client/tests/index.test.js (hunk 5 above)
Comment (MadeByMike) [reply packages/server-side-graphql-client/tests/index.test.js:22]: I'm open to condensing it but don't want to do it :)

File: packages/server-side-graphql-client/package.json (hunk 4 above)
Comment (MadeByMike) [reply packages/server-side-graphql-client/package.json:3]: Almost makes sense if you squint... but agree not a top-shelf sentence.

Comment (MadeByMike) [inline packages/server-side-graphql-client/package.json:3]:
```suggestion
  "description": "A library for running server-side graphQL queries and mutations in Keystone.",
```

File: packages/server-side-graphql-client/README.md (hunk 8 above)
Comment (MadeByMike) [reply packages/server-side-graphql-client/README.md:72]: Yep bad merge, looks like Aman and I updated the same thing

**Comment by MadeByMike** [comment]
> LGTM 👍

I'm still arguing that `extraContext` is the wrong API and I think I have convinced Tim. He's going to update it then it GTG.

**Comment by JedWatson** [comment]
I'm a bit late to the party but before we release this, `listName` needs to be `listKey`

**Pull request merged**